libr/core/cmd_type.c

```diff
@@ -334,6 +334,78 @@ static void cmd_type_noreturn(RCore *core, const char *input) {
 	}
 }
 
+static int calculate_type_size(Sdb *sdb_types, char *name) {
+	char *type = NULL;
+	if (!sdb_types || !name || !sdb_exists (sdb_types, name) || !(type = sdb_get (sdb_types, name, 0))) {
+		return 0;
+	}
+	char *type_name = r_str_newf ("%s.%s", type, name);
+	char *type_name_size = r_str_newf ("%s.%s.%s", type, name, "size");
+	int size = 0;
+	if (sdb_exists (sdb_types, type_name_size)) { // size already exists so stop recursion
+		char *v = sdb_get (sdb_types, type_name_size, 0);
+		size = sdb_atoi (v);
+		free (v);
+	} else if (!strcmp (type, "struct") || !strcmp (type, "union")) {
+		int idx;
+		char *v = NULL;
+		//loop through the contents to find individual sizes
+		for (idx = 0; (v = sdb_array_get (sdb_types, type_name, idx, NULL)); ++idx) {
+			char *sub_name = sdb_array_get (sdb_types, sdb_fmt ("%s.%s", type_name, v), 0, NULL);
+			char *ptr = NULL;
+			int sz = 0;
+			if (strchr (sub_name, '*')) {
+				// It is a pointer variable so get its size;
+				sz = calculate_type_size (sdb_types, "void *");
+			} else if ((ptr = strchr (sub_name, ' '))) {
+				sz = calculate_type_size (sdb_types, ptr + 1);
+			} else {
+				sz = calculate_type_size (sdb_types, sub_name);
+			}
+			// count stores the number of elements in case of an array
+			int count = sdb_array_get_num (sdb_types, sdb_fmt ("%s.%s", type_name, v), 2, NULL);
+			if (count) {
+				sz *= count;
+			}
+			if (!strcmp (type, "struct")) {
+				// size of struct is sum of sizes of its elements
+				size += sz;
+			} else {
+				// size of union is max of the sizes of its elements
+				size = sz > size ? sz : size;
+			}
+			free (sub_name);
+			free (v);
+		}
+	}
+	sdb_set (sdb_types, type_name_size, sdb_fmt ("%d", size), 0); // store the size for fast access
+	free (type);
+	free (type_name);
+	free (type_name_size);
+	return size;
+}
+
+R_API void save_parsed_type_size(RCore *core, const char *parsed) {
+	if (!core || !core->anal || !parsed) {
+		return;
+	}
+	char *str = strdup (parsed);
+	if (str) {
+		char *ptr = NULL;
+		int offset = 0;
+		while ((ptr = strstr (str + offset, "=struct\n")) || (ptr = strstr (str + offset, "=union\n"))) {
+			*ptr = 0;
+			char *name = ptr;
+			while (name - 1 >= str && *(name - 1) != '\n' && *(name - 1) != 0) {
```

Comment:
!= 0 is implicit, the *name-1) things are scary, better not to do negative detlas with pointers its a bad practice

Reply:
Okay! Thanks. Will keep the negative deltas in mind. For now I changed this to:
```
while (name - 1 >= str && *(name - 1) != '\n') {
	name--;
}
```
Ensuring `name - 1 >= str` before the other check should be safe for now.